### PR TITLE
[WIP] Add additional regressors

### PR DESCRIPTION
**Pull request by cml54**
added four regressors to skll
- BayesianRidge
- LARS
- RidgeCV

### CODE REVIEW

File: skll/learner.py [hunk 1]
```diff
@@ -40,9 +40,9 @@
 # AdditiveChi2Sampler is used indirectly, so ignore linting message
 from sklearn.kernel_approximation import (AdditiveChi2Sampler, Nystroem,
                                           RBFSampler, SkewedChi2Sampler)
-from sklearn.linear_model import (ElasticNet, Lasso, LinearRegression,
+from sklearn.linear_model import (ElasticNet, Lasso, Lars, LinearRegression,
                                   LogisticRegression, Ridge, SGDClassifier,
-                                  SGDRegressor)
+                                  SGDRegressor, BayesianRidge, RidgeCV)
```
Comment (desilinguist) [inline skll/learner.py]: Please sort these alphabetically since that is the convention for imports.

File: skll/learner.py [hunk 2]
```diff
@@ -107,8 +109,14 @@
...
-                          'gamma': [0.01, 0.1, 1.0, 10.0, 100.0]}]}
-
+                          'gamma': [0.01, 0.1, 1.0, 10.0, 100.0]}],
+                        RidgeCV:
+                        [{'alphas': [(0.01, 0.1, 1.0, 10.0, 100.0)]}],
+                        BayesianRidge:
+                        [{'alpha_1': [1.e-4, 1.e-5, 1.e-6, 1.e-7, 1.e-8],
+                          'alpha_2': [1.e-4, 1.e-5, 1.e-6, 1.e-7, 1.e-8],
+                          'lambda_1': [1.e-4, 1.e-5, 1.e-6, 1.e-7, 1.e-8],
+                          'lambda_2': [1.e-4, 1.e-5, 1.e-6, 1.e-7, 1.e-8]}]}
```
Comment (desilinguist) [inline skll/learner.py]: How did you come up with these values for these regressors? Are they just guesses or are they the sorts of values people try? You might have to read some papers to make sure.

File: skll/learner.py [hunk 3]
```diff
@@ -144,7 +152,10 @@
                                   'lwk_off_by_one',
                                   'qwk_off_by_one'])
 
-_REQUIRES_DENSE = (GradientBoostingClassifier, GradientBoostingRegressor)
+
+_REQUIRES_DENSE = (GradientBoostingClassifier, GradientBoostingRegressor,
+                   BayesianRidge, Lars)
+
```
Comment (desilinguist) [inline skll/learner.py]: Alphabetical, please.

File: tests/test_metrics.py [hunk 4]
```diff
@@ -136,9 +136,11 @@ def check_invalid_regr_grid_obj_func(learner_name, grid_objective_function):
 def test_invalid_grid_obj_func():
     for model in ['AdaBoostRegressor', 'DecisionTreeRegressor',
                   'ElasticNet', 'GradientBoostingRegressor',
-                  'KNeighborsRegressor', 'Lasso',
+                  'KNeighborsRegressor', 'Lasso', 'Lars',
                   'LinearRegression', 'RandomForestRegressor',
-                  'Ridge', 'LinearSVR', 'SVR', 'SGDRegressor']:
+                  'Ridge', 'LinearSVR', 'SVR', 'SGDRegressor',
+                  'RidgeCV', 'BayesianRidge']:
+
```
Comment (desilinguist) [inline tests/test_metrics.py]: Alphabetical, again.

File: tests/test_regression.py [hunk 5]
```diff
@@ -130,8 +130,10 @@ def check_rescaling(name):
 
 
 def test_rescaling():
-    for regressor_name in ['ElasticNet', 'Lasso', 'LinearRegression', 'Ridge',
-                           'LinearSVR', 'SVR', 'SGDRegressor']:
+
+    for regressor_name in ['ElasticNet', 'Lasso', 'Lars', 'LinearRegression',
+                           'Ridge', 'LinearSVR', 'SVR', 'SGDRegressor',
+                           'RidgeCV', 'BayesianRidge']:
         yield check_rescaling, regressor_name
```
Comment (desilinguist) [inline tests/test_regression.py]: Alphabetical.

**Comment by desilinguist** [comment]
You also need to update the documentation under `doc` and provide descriptions of the new learners as well as the tuning grids.

File: skll/learner.py (hunk 3 above)
Comment (cml54) [inline skll/learner.py]: Done.

File: tests/test_metrics.py (hunk 4 above)
Comment (cml54) [inline tests/test_metrics.py]: Done.

File: tests/test_regression.py (hunk 5 above)
Comment (cml54) [inline tests/test_regression.py]: done.

**Comment by desilinguist** [comment]
@cml54 and @benleong-ets do you know if you will have time to work on this again in the next week or so?

**Comment by benleong-ets** [comment]
Hi @desilinguist, sorry for the delay - this week is a bit tight for me with some other deadlines. I think I can put a few hours into it next week. Just to clarify: the grid search values I put in are orders of magnitude larger/smaller than the default, which I observe are consistent with previously implemented regressors. But, I would need to do a bit of literature research to see if those are really the case for the one I implemented i.e. Lars. for further validation.

**Comment by desilinguist** [comment]
@benleong-ets that's totally fine. There's no urgency about adding these to SKLL so we can do it in the next release. Thanks!

**Comment by benleong-ets** [comment]
Sounds good. Thanks!

**Comment by desilinguist** [comment]
@cml54 and @benleong-ets I am assuming this PR is still not done?

**Comment by cml54** [comment]
@desilinguist 

I didn't have time to address it. I will find time to work on it in two weeks. Sorry for the delay.

**Comment by benleong-ets** [comment]
Sorry @desilinguist, it is still not done yet. Let me try to get it done asap within the next few weeks.

**Comment by benleong-ets** [comment]
Are we planning for another release soon?

**Comment by benleong-ets** [comment]
Updated earlier comment (on 02/18)
@desilinguist, according to what understand from the literature, it seems that the grid search values for  n_nonzero_coefs  (Lars) should be cover the extreme values explored in the forward selection algorithm i.e. we can be highly discriminative and decide that only the variable with highest correlation to the residual should be included in the model, or adopt a fully "democratic" approach where every variable should be included, or adopt an approach where a subset of the variables are included. So, I would modify the search values from {'n_nonzero_coefs': [5, 50, 500, 5000, 50000]} to  [1, 50, 500, 5000, np.inf]. 500 is the default, while 1 and np.inf are the ends of the search. Does this make sense?

Useful links:
http://statweb.stanford.edu/~tibs/lasso/simple.html
https://www.quora.com/What-is-Least-Angle-Regression-and-when-should-it-be-used

**Comment by benleong-ets** [comment]
@desilinguist, I have added and tested most of the regressors, except for `lars_path`, `lasso_path` and `lasso_stability_path`, which are method-based instead of class-based. For these, I am not sure how to add them without major changes to the current structure of the code. 

There is also an issue during testing after implementing `LassoLars` that I could not get around by tuning the parameters. I basically tried `alpha`, `eps`, `fit_intercept`, `positive`, `max_iter` etc, but still get the following errors when testing in `test_regression.test_linear_models`. Similar errors happened to `ElasticNetCV`, but it was solved by adjusting for `eps` using grid search. 

```
======================================================================
FAIL: test_regression.test_linear_models('LassoLars', False, False)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/cleong/anaconda/envs/skll/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/Users/cleong/skll/tests/test_regression.py", line 184, in check_linear_models
    eq_(learned_w, given_w)
nose.proxy.AssertionError: 5 != 37
-------------------- >> begin captured logging << --------------------
skll.learner: WARNING: Training data will be shuffled to randomize grid search folds.  Shuffling may yield different results compared to scikit-learn.
--------------------- >> end captured logging << ---------------------

======================================================================
FAIL: test_regression.test_linear_models('LassoLars', False, True)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/cleong/anaconda/envs/skll/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/Users/cleong/skll/tests/test_regression.py", line 184, in check_linear_models
    eq_(learned_w, given_w)
nose.proxy.AssertionError: 5 != 37
-------------------- >> begin captured logging << --------------------
skll.learner: WARNING: Training data will be shuffled to randomize grid search folds.  Shuffling may yield different results compared to scikit-learn.
--------------------- >> end captured logging << ---------------------

======================================================================
FAIL: test_regression.test_linear_models('LassoLars', True, False)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/cleong/anaconda/envs/skll/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/Users/cleong/skll/tests/test_regression.py", line 196, in check_linear_models
    assert_greater(cor, expected_cor_range[0])
nose.proxy.AssertionError: 0.59224644892445055 not greater than 0.7
-------------------- >> begin captured logging << --------------------
skll.learner: WARNING: Training data will be shuffled to randomize grid search folds.  Shuffling may yield different results compared to scikit-learn.
--------------------- >> end captured logging << ---------------------

======================================================================
FAIL: test_regression.test_linear_models('LassoLars', True, True)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/cleong/anaconda/envs/skll/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/Users/cleong/skll/tests/test_regression.py", line 196, in check_linear_models
    assert_greater(cor, expected_cor_range[0])
nose.proxy.AssertionError: 0.59224644892445055 not greater than 0.7
-------------------- >> begin captured logging << --------------------
skll.learner: WARNING: Training data will be shuffled to randomize grid search folds.  Shuffling may yield different results compared to scikit-learn.
--------------------- >> end captured logging << ---------------------
```

There are also weird cases where sometimes local testing passed but Travis testing would fail, and these appear to be random. I tried to narrow down to the cause where a specific parameter for a base linear model would be applied to a derived linear model, but, because such a param is not listed explicitly for the derived model (instead, it is passed in as part of the base object), an error was thrown. For example, specifying grid values for the `n_nonzero_coefs` param for `Lars` resulted in an `invalid parameter` error for `LassoLarsIC`. Here is the [error](https://travis-ci.org/EducationalTestingService/skll/jobs/134293803#L4450). I have resolved these type of errors by restricting grid search parameters to those common between the base/derived models. 

There are still some cleaning to be done. But, when you get a chance, please review the changes and let me know if you agree to the way these issues are handled.

Thank you!

**Comment by benbuleong** [comment]
@desilinguist: Hi Nitin, I am wondering if you could get the chance to look at the issue I posted above in the next week or so?

**Comment by desilinguist** [comment]
@benbuleong, I don't think I will be able to given what's on my plate now. Probably next month. Sorry.

**Pull request closed without merging**